sklearn/model_selection/tests/test_validation.py

```diff
@@ -936,6 +937,138 @@ def test_cross_val_predict_with_method():
         assert_array_almost_equal(expected_predictions, predictions)
 
 
+def test_cross_val_predict_class_subset():
+    iris = load_iris()
+    X, y = iris.data, iris.target
+    X, y = shuffle(X, y, random_state=0)
+    classes = len(set(y))
+
+    # Modifies the dataset so that in a particular fold, produced by
+    # kfold.split, the training set is composed of only 2 classes instead of 3
+    y[:50] = 0
+    y[50:100] = 1
+    X[:100], y[:100] = shuffle(X[:100], y[:100], random_state=0)
+
+    kfold = KFold(len(iris.target))
+
+    methods = ['predict_proba', 'predict_log_proba']
+    for method in methods:
+        est = LogisticRegression()
+
+        predictions = cross_val_predict(est, X, y, method=method)
+        assert_equal(len(predictions), len(y))
+
+        expected_predictions = np.zeros([len(y), classes])
+        func = getattr(est, method)
+
+        # Naive loop (should be same as cross_val_predict):
+        for train, test in kfold.split(X, y):
+            est.fit(X[train], y[train])
+            expected_predictions_ = func(X[test])
+            # To avoid 2 dimensional indexing
+            exp_pred_test = np.zeros((len(test), classes))
+            if method is 'decision_function' and len(est.classes_) == 2:
+                exp_pred_test[:, est.classes_[-1]] = expected_predictions_
+            else:
+                exp_pred_test[:, est.classes_] = expected_predictions_
+            expected_predictions[test] = exp_pred_test
+
+        predictions = cross_val_predict(est, X, y, method=method,
+                                        cv=kfold)
+        assert_array_almost_equal(expected_predictions, predictions)
+
+
+def test_cross_val_predict_different_label_types():
+    iris = load_iris()
+    X, y = iris.data, iris.target
+    X, y = shuffle(X, y, random_state=0)
+    classes = len(set(y))
+
+    # unordered integer labels
+    y_ = np.empty(y.shape)
+    for i, c in zip(range(3), [-1, 4, 6]):
+        y_[np.where(y == i)] = c
+
+    # Modifies the dataset so that in a particular fold, produced by
+    # kfold.split, the training set is composed of only 2 classes instead of 3
+    y_[:50] = -1
+    y_[50:100] = 4
+    X[:100], y_[:100] = shuffle(X[:100], y_[:100], random_state=0)
+
+    # string labels
+    y_str = np.empty(y.shape, dtype=object)
+    for i, c in zip(range(3), iris.target_names):
+        y_str[np.where(y == i)] = c
+
+    # Modifies the dataset so that in a particular fold, produced by
+    # kfold.split, the training set is composed of only 2 classes instead of 3
```

Comment:
Is it not appropriate to just check in `test_cross_val_predict_with_method` whether alternative representations of y (including `y + 1`, `y - 2`, `y.astype('U')`) yield the same predictions?